### ISSUE TITLE
parser improvement: span for "expected one of ..." should include preceding token

### ISSUE THREAD

**Issue by pnkfelix** [rust-lang/rust#25177]
This is just a User Experience request for our parser's error messages.

Consider the following code ([playpen](http://is.gd/GZs71v))

``` rust
fn main()
{
    let ab = {
        let thing = match (1, 2) {
            (_, _) => println!("Hello"),
        }
        (3, 4)
    };
}
```

This prints the error message

```
<anon>:8:5: 8:6 error: expected one of `.`, `;`, or an operator, found `}`
<anon>:8     };
             ^
```

I spent a little while staring at this confused, saying to myself "but there _is_ a semicolon there, after the `}`" ... of course, the message was 100% correct in its content.

The question is: What can we do to provide better feedback in a case like this (where the true error lies earlier in the stream).

One idea I had was that it might help to increase the span to include the preceding token, i.e. to make the message indicate "I found a `}` after the `)` in `(3,4)`, expected one of ..."

Another idea: While parsing, track stylistic anti-patterns like whitespace separating a function expression from the parentheses for its arguments in a call-expression, and then report them as part of the error report.

**Comment by XuefengWu** [comment]
how about check `let` statement **must** end with semicolon?

**Comment by pnkfelix** [comment]
@XuefengWu that *is* what it is complaining about -- all `let`-statements *must* end with semi-colons.

The problem is that a single expression can cross multiple lines, and you cannot tell if it might be a function call at the point where you might naively expect it to complain.

To be clear about this: the problem is that the parser *cannot complain* at this earlier point in a parse of a slight variation on the above code:

```rust
        let thing = match (1, 2) {
            (_, _) => foo,
        }
        (3, 4);
//       ^
//       | : Parser is here, it has just read the `(`.
```

because *this* (variant) code actually should parse: it is going to evaluate the `match` to `foo`, which could have function type `foo: fn(i32, i32) -> Thing`.

So the parser keeps reading over the `(3, 4)`.

And in the original code, after reading the `(3, 4)`, it then sees the `}`, and *then* complains, because it knows *that* (the closing curly brace) is surely incorrect.

This is why I was suggesting things like tracking stylistic anti-patterns: even though things like:
```rust
    // (Rust is not Javascript; there is no automatic-semicolon insertion.)
    let thing = foo
    (3, 4);
```
are *legal*, they are also stylistic anti-patterns, and reporting them in the event of a parse error might be helpful.  Maybe.

**Comment by steveklabnik** [comment]
Triage: no change

**Comment by Mark-Simulacrum** [comment]
It seems that we give a much better span here today. Closing.

```
error: expected one of `.`, `;`, `?`, or an operator, found `}`
 --> test.rs:8:5
  |
7 |         (3, 4)
  |               - expected one of `.`, `;`, `?`, or an operator here
8 |     };
  |     ^ unexpected token

error: expected item, found `;`
 --> test.rs:8:6
  |
8 |     };
  |      ^

error: aborting due to previous error(s)
```
